### PR TITLE
Small HA tracker cleanup

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This PR does some small code "cleanup" in HA tracker (metrics as fields, logger), and tests (removes usage of `mtime`, uses `test.Poll`). In addition to that, there is new integration test which doesn't verify behaviour, only logs it -- it shows differences in WatchPrefix and Delete behaviour between Consul (existing keys are returned via WatchPrefix, Delete doesn't send "notification" to WatchPrefix) and etcd (existing keys are not returned via WatchPrefix, but Delete "sends" notification).

**Checklist**
- [x] Tests updated
- [na] Documentation added
- [na] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: integration/kv_test.go [hunk 1]
```diff
@@ -20,18 +22,95 @@ import (
...
+		wg.Add(1)
+		go w.watch(ctx, wg, client)
+
+		err = client.CAS(context.Background(), "key-to-delete", func(in interface{}) (out interface{}, retry bool, err error) {
+			return "value-to-delete", false, nil
+		})
+		require.NoError(t, err, "object could not be created")
+
+		// Now delete it
+		err = client.Delete(context.Background(), "key-to-delete")
```
Comment (pracucci) [inline integration/kv_test.go:95]: Should we add a sleep between creating the `key-to-delete` and deleting it? I'm wondering what happens if the two operations happens very close each other: the watcher may not get any notification (depending on how the backend is implemented).

File: pkg/distributor/distributor_test.go [hunk 2]
```diff
@@ -524,15 +525,15 @@ func TestDistributor_PushHAInstances(t *testing.T) {
 						KVStore:         kv.Config{Mock: mock},
 						UpdateTimeout:   100 * time.Millisecond,
 						FailoverTimeout: time.Second,
-					}, trackerLimits{maxClusters: 100}, nil)
+					}, trackerLimits{maxClusters: 100}, nil, util_log.Logger)
```
[the remark this answers is not in the document]
Comment (jtlisi) [reply pkg/distributor/distributor_test.go:528]: +1

[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/distributor/distributor_test.go:528]: Sure, although I've only found tcp_transport.go using nop logger. 🤔

File: pkg/distributor/ha_tracker_test.go [hunk 3]
```diff
@@ -388,36 +352,28 @@ func TestCheckReplicaMultiUser(t *testing.T) {
...
 	assert.NoError(t, err)
+	checkReplicaTimestamp(t, time.Second, c, user, cluster, replica, now)
 
 	// Write the first time for user 2.
-	err = c.checkReplica(ctxUser2, user, cluster, replica)
-	assert.NoError(t, err)
-	ctx, cancel = context.WithTimeout(ctxUser2, time.Second)
-	err = checkReplicaTimestamp(ctx, c, user, cluster, replica, start)
-	cancel()
+	err = c.checkReplica(context.Background(), user, cluster, replica, now)
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/distributor/ha_tracker_test.go:368]: You're right, good catch. I'm bit puzzled as well... I don't think old test was really doing what the test name says. I'll update it to do so.

File: pkg/distributor/ha_tracker.go [hunk 4]
```diff
@@ -83,6 +59,11 @@ type haTracker struct {
 	electedLock sync.RWMutex
 	elected     map[string]ReplicaDesc // Replicas we are accepting samples from. Key = "user/cluster".
 	clusters    map[string]int         // Number of clusters with elected replicas that a single user has. Key = user.
+
+	electedReplicaChanges         *prometheus.CounterVec
+	electedReplicaTimestamp       *prometheus.GaugeVec
+	electedReplicaPropagationTime prometheus.Histogram
+	kvCASCalls                    *prometheus.CounterVec
```
Comment (jtlisi) [inline pkg/distributor/ha_tracker.go:66]: **[nit,non-blocking]:** Would it be possible to move the `haTracker` definition to just above the `newClusterTracker` function? It makes it a bit easier to read when you see the metric definitions and fields near one another.

File: integration/kv_test.go [hunk 5]
```diff
@@ -20,18 +22,95 @@ import (
...
+		require.NoError(t, err)
+
+		// Stop the watcher
+		cancel()
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply integration/kv_test.go:99]: You're right, and it's fixed in #3809, but let's add a wait here.

File: integration/kv_test.go (hunk 1 above)
Comment (pstibrany) [reply integration/kv_test.go:95]: OK, let's do that.

File: pkg/distributor/ha_tracker.go (hunk 4 above)
Comment (pstibrany) [reply pkg/distributor/ha_tracker.go:66]: 👍 I did that, and also renamed `newClusterTracker` to `newHATracker`.

Comment (jtlisi) [reply pkg/distributor/ha_tracker.go:66]: 👍

**Comment by pstibrany** [comment]
Thanks for reviews, I've addressed your feedbacks, please take a look again.

File: pkg/distributor/distributor_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/distributor/distributor_test.go]: My bad, I was searching with case-sensitive search 🤦

File: pkg/distributor/ha_tracker_test.go [hunk 6]
```diff
@@ -132,68 +112,64 @@ func TestHATrackerConfig_Validate(t *testing.T) {
...
 	mock := kv.PrefixClient(consul.NewInMemoryClient(codec), "prefix")
-	c, err := newClusterTracker(HATrackerConfig{
+	c, err := newHATracker(HATrackerConfig{
 		EnableHATracker:        true,
 		KVStore:                kv.Config{Mock: mock},
 		UpdateTimeout:          time.Millisecond,
 		UpdateTimeoutJitterMax: 0,
 		FailoverTimeout:        time.Millisecond * 2,
-	}, trackerLimits{maxClusters: 100}, nil)
+	}, trackerLimits{maxClusters: 100}, nil, util_log.Logger)
```
Comment (jtlisi) [inline pkg/distributor/ha_tracker_test.go:124]:
```suggestion
	}, trackerLimits{maxClusters: 100}, nil, log.NewNopLogger())
```

File: pkg/distributor/ha_tracker_test.go [hunk 7]
```diff
@@ -132,68 +112,64 @@ func TestHATrackerConfig_Validate(t *testing.T) {
...
 
-	c, err := newClusterTracker(HATrackerConfig{
+	c, err := newHATracker(HATrackerConfig{
 		EnableHATracker:        true,
 		KVStore:                kv.Config{Store: "inmemory"},
 		UpdateTimeout:          100 * time.Millisecond,
 		UpdateTimeoutJitterMax: 0,
 		FailoverTimeout:        time.Second,
-	}, trackerLimits{maxClusters: 100}, nil)
+	}, trackerLimits{maxClusters: 100}, nil, util_log.Logger)
```
Comment (jtlisi) [inline pkg/distributor/ha_tracker_test.go:149]:
```suggestion
	}, trackerLimits{maxClusters: 100}, nil, log.NewNopLogger())
```

File: pkg/distributor/ha_tracker_test.go [hunk 8]
```diff
@@ -202,33 +178,35 @@ func TestCheckReplicaMultiCluster(t *testing.T) {
...
 	reg := prometheus.NewPedanticRegistry()
-	c, err := newClusterTracker(HATrackerConfig{
+	c, err := newHATracker(HATrackerConfig{
 		EnableHATracker:        true,
 		KVStore:                kv.Config{Store: "inmemory"},
 		UpdateTimeout:          100 * time.Millisecond,
 		UpdateTimeoutJitterMax: 0,
 		FailoverTimeout:        time.Second,
-	}, trackerLimits{maxClusters: 100}, reg)
+	}, trackerLimits{maxClusters: 100}, reg, util_log.Logger)
```
Comment (jtlisi) [inline pkg/distributor/ha_tracker_test.go:187]:
```suggestion
	}, trackerLimits{maxClusters: 100}, reg, log.NewNopLogger())
```

File: pkg/distributor/ha_tracker_test.go [hunk 9]
```diff
@@ -246,56 +224,57 @@ func TestCheckReplicaMultiCluster(t *testing.T) {
...
```
Comment (jtlisi) [inline pkg/distributor/ha_tracker_test.go:237]:
```suggestion
	}, trackerLimits{maxClusters: 100}, reg, log.NewNopLogger())
```

File: pkg/distributor/ha_tracker_test.go [hunk 10]
```diff
@@ -314,110 +293,88 @@ func TestCheckReplicaMultiClusterTimeout(t *testing.T) {
...
 	mock := kv.PrefixClient(consul.NewInMemoryClient(codec), "prefix")
-	c, err := newClusterTracker(HATrackerConfig{
+	c, err := newHATracker(HATrackerConfig{
 		EnableHATracker:        true,
 		KVStore:                kv.Config{Mock: mock},
 		UpdateTimeout:          time.Second,
 		UpdateTimeoutJitterMax: 0,
 		FailoverTimeout:        time.Second,
-	}, trackerLimits{maxClusters: 100}, nil)
+	}, trackerLimits{maxClusters: 100}, nil, util_log.Logger)
```
Comment (jtlisi) [inline pkg/distributor/ha_tracker_test.go:308]:
```suggestion
	}, trackerLimits{maxClusters: 100}, nil, log.NewNopLogger())
```

File: pkg/distributor/ha_tracker_test.go [hunk 11]
```diff
@@ -314,110 +293,88 @@ func TestCheckReplicaMultiClusterTimeout(t *testing.T) {
...
 	mock := kv.PrefixClient(consul.NewInMemoryClient(codec), "prefix")
-	c, err := newClusterTracker(HATrackerConfig{
+	c, err := newHATracker(HATrackerConfig{
 		EnableHATracker:        true,
 		KVStore:                kv.Config{Mock: mock},
 		UpdateTimeout:          100 * time.Millisecond,
 		UpdateTimeoutJitterMax: 0,
 		FailoverTimeout:        time.Second,
-	}, trackerLimits{maxClusters: 100}, nil)
+	}, trackerLimits{maxClusters: 100}, nil, util_log.Logger)
```
Comment (jtlisi) [inline pkg/distributor/ha_tracker_test.go:354]:
```suggestion
	}, trackerLimits{maxClusters: 100}, nil, log.NewNopLogger())
```

File: pkg/distributor/ha_tracker_test.go [hunk 12]
```diff
@@ -465,40 +422,36 @@ func TestCheckReplicaUpdateTimeoutJitter(t *testing.T) {
...
 			mock := kv.PrefixClient(consul.NewInMemoryClient(codec), "prefix")
-			c, err := newClusterTracker(HATrackerConfig{
+			c, err := newHATracker(HATrackerConfig{
 				EnableHATracker:        true,
 				KVStore:                kv.Config{Mock: mock},
 				UpdateTimeout:          testData.updateTimeout,
 				UpdateTimeoutJitterMax: 0,
 				FailoverTimeout:        time.Second,
-			}, trackerLimits{maxClusters: 100}, nil)
+			}, trackerLimits{maxClusters: 100}, nil, util_log.Logger)
```
Comment (jtlisi) [inline pkg/distributor/ha_tracker_test.go:431]:
```suggestion
			}, trackerLimits{maxClusters: 100}, nil, log.NewNopLogger())
```

File: pkg/distributor/ha_tracker_test.go [hunk 13]
```diff
@@ -561,38 +514,38 @@ func TestHATrackerConfig_ShouldCustomizePrefixDefaultValue(t *testing.T) {
...
 
-	t1, err := newClusterTracker(HATrackerConfig{
+	t1, err := newHATracker(HATrackerConfig{
 		EnableHATracker:        true,
 		KVStore:                kv.Config{Mock: mock},
 		UpdateTimeout:          time.Second,
 		UpdateTimeoutJitterMax: 0,
 		FailoverTimeout:        time.Second,
-	}, limits, nil)
+	}, limits, nil, util_log.Logger)
```
Comment (jtlisi) [inline pkg/distributor/ha_tracker_test.go:529]:
```suggestion
	}, limits, nil, log.NewNopLogger())
```

File: pkg/distributor/ha_tracker_test.go [hunk 14]
```diff
@@ -643,27 +596,27 @@ func (l trackerLimits) MaxHAClusters(_ string) int {
 }
 
 func TestHATracker_MetricsCleanup(t *testing.T) {
-	electedReplicaChanges.Reset()
-	electedReplicaTimestamp.Reset()
-	kvCASCalls.Reset()
+	reg := prometheus.NewPedanticRegistry()
+	tr, err := newHATracker(HATrackerConfig{EnableHATracker: false}, nil, reg, util_log.Logger)
```
Comment (jtlisi) [inline pkg/distributor/ha_tracker_test.go:600]:
```suggestion
	tr, err := newHATracker(HATrackerConfig{EnableHATracker: false}, nil, reg, log.NewNopLogger())
```

**Review by jtlisi — APPROVED** [review]
LGTM

Modulo `%s/util_log.Logger/log.NewNopLogger()/g`

**Review by pracucci — APPROVED** [review]
LGTM (modulo the comments about the nooplogger in tests).

I haven't checked, but if the `pkg/distributor` is not using the global logger anymore than please add it to the failling "rule" in Makefile:

```
	faillint -paths "github.com/cortexproject/cortex/pkg/util/log.{Logger}" \
		./pkg/ingester/... \
		./pkg/flusher/... \
		./pkg/distributor/... 
```

**Comment by pstibrany** [comment]
> LGTM (modulo the comments about the nooplogger in tests).
> 
> I haven't checked, but if the `pkg/distributor` is not using the global logger anymore than please add it to the failling "rule" in Makefile:

We're not quite there yet due to logger in `(*RingConfig).RegisterFlags` but I would leave that for another PR. Distributor itself is now global-logger free though.

**Pull request merged**